src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -358,7 +363,7 @@ object RuleUtils {
       // Currently, BucketUnion does not keep the sort order within a bucket.
       BucketUnion(Seq(indexPlan, shuffled), bucketSpec)
     } else {
-      indexPlan
+      Union(Seq(indexPlan, complementIndexPlan))
```

Comment:
There is a comment earlier in this file:

```
    // Though BucketUnion (using BucketSpec and on-the-fly Shuffle) is used to merge
    // them for now, these plans will be optimized with a Union operator at a later time
    // (see #145).
```

Don't forget to remove this comment when you are merging this PR.